https://www.datamanim.com/dataset/03_dataq/typeone.html#id3

- 유튜브 인기동영상 데이터 (O)
- 유튜브 공범컨텐츠 동영상 데이터 (O)
- 월드컵 출전선수 골기록 데이터 (O)
- 서울시 따릉이 이용정보 데이터 (O)
- 전세계 행복도 지표 데이터 (O)
- 지역구 에너지 소비량 데이터 (O)
- 포켓몬 정보 데이터 (O)
- 대한민국 체력장 데이터
- 기온 강수량 데이터
- 서비스 이탈예측 데이터
- 성인 건강검진 데이터
- 자동차 보험가입 예측데이터
- 핸드폰 가격 예측데이터
- 비행탑승 경험 만족도 데이터
- 수질 음용성 여부 데이터
- 의료 비용 예측 데이터
- 킹카운티 주거지 가격예측문제 데이터
- 대학원 입학가능성 데이터
- 레드 와인 퀄리티 예측 데이터
- 약물 분류 데이터
- 사기회사 분류 데이터
- 센서데이터 동작유형 분류 데이터
- 현대 차량 가격 분류문제 데이터
- 당뇨여부판단 데이터
- 넷플릭스 주식 데이터
- nba 선수 능력치 데이터


# 유튜브 인기동영상 데이터

데이터 설명 : 유튜브 데일리 인기동영상 (한국)

In [1]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/youtube.csv",index_col=0)
df.head()

title             channelTitle  \
0                                          [신병] 물자창고                      장삐쭈   
1   RAIN(비) - 나로 바꾸자 Switch to me (duet with JYP) MV  RAIN's Official Channel   
2       2020년 제야의 종 온라인 타종행사 | 보신각 현장 행사는 진행하지 않습니다.              서울시 · Seoul   
3                                       고기남자의 칠면조 파티             고기남자 MeatMan   
4  골목 3mc를 분노하게 만든 마음고생이 심했을 공릉 백반집 사장님의 푸념?! [예능...                    스브스밥집   

   categoryId  view_count  likes  dislikes  comment_count  \
0          23     1893473  38249       730           8595   
1          10     2600864      0         0          20129   
2          29      347049   3564       120            178   
3          26      528458  15372       280           3470   
4          24      494904   3918       111           3142   

                  channelId trending_date2  
0  UChbE5OZQ6dRHECsX0tEPEZQ     2021-01-01  
1  UCxXgIeE5hxWxHG6dz9Scg2w     2021-01-01  
2  UCZUPZW5idAxYp-Asj__lVAA     2021-01-01  
3  UCT3CumbFIJiW33uq0UI3zlg     2021-01-01  
4  UCdWgRSfttvDucq4ApcCg5Mw     2021-01-01

Question 1

인기동영상 제작횟수가 많은 채널 상위 10개명을 출력하라 (날짜기준, 중복포함)

In [2]:
# 인기 동영상 제작 횟수 상위 10개 채널 id
df.channelId.value_counts().head(10)

channelId
UCszFjh7CEfwDb7UUGb4RzCQ    372
UClzB2iZ5jPoTNz0S-QU6Wiw    318
UCtm_QoN2SIxwCE-59shX7Qg    318
UCaKod3X1Tn4c7Ci0iUKcvzQ    313
UCiwQRG2sCcfjKkgxMEdJGPg    293
UChbE5OZQ6dRHECsX0tEPEZQ    283
UCLkAepWjdylmXSltofFvsYQ    275
UCQ2O-iftmnlfrBuNsUUTofQ    274
UCtfriFvIe5gIe8wRWb8UTMg    258
UCRuSxVu4iqTK5kCh90ntAgA    255
Name: count, dtype: int64

In [3]:
df[df.channelId.isin(df.channelId.value_counts().head(10).index)].channelTitle.unique()

# ['장삐쭈', '총몇명', '파뿌리', '짤툰', '런닝맨 - 스브스 공식 채널',
# '엠뚜루마뚜루 : MBC 공식 종합 채널', 'SPOTV', '채널 십오야', '이과장', 'BANGTANTV']

array(['장삐쭈', '총몇명', '파뿌리', '짤툰', '런닝맨 - 스브스 공식 채널',
       '엠뚜루마뚜루 : MBC 공식 종합 채널', 'SPOTV', '채널 십오야', '이과장', 'BANGTANTV'],
      dtype=object)

Question 2

논란으로 인기동영상이 된 케이스를 확인하고 싶다. dislikes수가 like 수보다 높은 동영상을 제작한 채널을 모두 출력하라

In [4]:
df[df['dislikes'] > df['likes']].channelTitle.unique()

# ['핫도그TV', 'ASMR 애정TV', '하얀트리HayanTree', '양팡 YangPang',
#  '철구형 (CHULTUBE)', '왜냐맨하우스', '(MUTUBE)와꾸대장봉준', '오메킴TV', '육지담',
#  'MapleStory_KR', 'ROAD FIGHTING CHAMPIONSHIP', '사나이 김기훈',
#  '나혼자산다 STUDIO', 'Gen.G esports']

array(['핫도그TV', 'ASMR 애정TV', '하얀트리HayanTree', '양팡 YangPang',
       '철구형 (CHULTUBE)', '왜냐맨하우스', '(MUTUBE)와꾸대장봉준', '오메킴TV', '육지담',
       'MapleStory_KR', 'ROAD FIGHTING CHAMPIONSHIP', '사나이 김기훈',
       '나혼자산다 STUDIO', 'Gen.G esports'], dtype=object)

Question 3

채널명을 바꾼 케이스가 있는지 확인하고 싶다. channelId의 경우 고유값이므로 이를 통해 채널명을 한번이라도 바꾼 채널의 갯수를 구하여라

In [5]:
# channelId에 channelTitle 개수
cnt = df[['channelTitle','channelId']].drop_duplicates().channelId.value_counts()
cnt

channelId
UCtQYhzJWFai1y30UIA3uwYQ    3
UC98TOxKQk4aLcx0EjIK0LkQ    3
UC4nEQI74rZkFKlUSwi0ur4A    2
UCn64fSqF2l4rkkjtQUpku2A    2
UCjO3ji06DxmTM1RrzLVwmkA    2
                           ..
UCuZu8NrpBG4WPXRi-hPBl-A    1
UCDrAR1OWC2MD4s0JLetN0MA    1
UCx2MVBsdTJ_sv0LiUpSQU2Q    1
UCu2R3TCLSxYV8pRyBGNOh2g    1
UCRCK5FCJtomQT3b88jXI_DA    1
Name: count, Length: 1770, dtype: int64

In [6]:
len(cnt[cnt>1])

# 71

71

Question 4

일요일에 인기있었던 영상들중 가장많은 영상 종류(categoryId)는 무엇인가?

In [7]:
df['trending_date2'] =  pd.to_datetime(df['trending_date2'])
df[df['trending_date2'].dt.day_name() == 'Sunday'].categoryId.value_counts().index[0]

# 24

24

Question 5

각 요일별 인기 영상들의 categoryId는 각각 몇개 씩인지 하나의 데이터 프레임으로 표현하라

In [8]:
# 각 요일별 categoryId 각각 몇개씩인가
a = df.groupby([df['trending_date2'].dt.day_name(),'categoryId'],as_index=False).size()
a

trending_date2  categoryId  size
0           Friday           1   243
1           Friday           2   120
2           Friday          10   833
3           Friday          15   187
4           Friday          17   633
..             ...         ...   ...
100      Wednesday          25   468
101      Wednesday          26   385
102      Wednesday          27   212
103      Wednesday          28   165
104      Wednesday          29    12

[105 rows x 3 columns]

In [9]:
a.pivot(index='categoryId',columns='trending_date2')

size                                                  
trending_date2 Friday Monday Saturday Sunday Thursday Tuesday Wednesday
categoryId                                                             
1                 243    263      255    274      246     257       234
2                 120    105      119     99      128     119       129
10                833    837      776    830      890     894       917
15                187    215      198    217      207     208       207
17                633    668      592    636      682     708       706
19                 90     92       87     91       92      89        85
20                283    298      296    289      282     285       291
22               1288   1373     1289   1337     1341    1375      1333
23                568    594      570    556      560     569       566
24               2976   3148     3066   3096     2954    3084      3090
25                444    453      422    437      470     452       468
26                369    378      364    363      375     394       385
27                183    205      183    199      194     194       212
28                171    160      173    167      166     161       165
29                 12     10       10      9       13      11        12

Question 6

댓글의 수로 (comment_count) 영상 반응에 대한 판단을 할 수 있다. viewcount대비 댓글수가 가장 높은 영상을 확인하라 (view_count값이 0인 경우는 제외한다)

In [10]:
# view_count값이 0인 경우 제외
a = df[df['view_count'] > 0].copy()
a['ratio'] = (a['comment_count']/a['view_count'])
a.sort_values('ratio',ascending=False).iloc[0].title

# 60분 동안 댓글이 달리지 않으면, 영상이 삭제됩니다. (챌린지)

'60분 동안 댓글이 달리지 않으면, 영상이 삭제됩니다. (챌린지)'

Question 7

댓글의 수로 (comment_count) 영상 반응에 대한 판단을 할 수 있다.viewcount대비 댓글수가 가장 낮은 영상을 확인하라 (view_counts, ratio값이 0인경우는 제외한다.)

In [13]:
a[a['ratio']!= 0].sort_values('ratio').iloc[0].title

# Join the BTS #PermissiontoDance Challenge only on YouTube #Shorts

'Join the BTS #PermissiontoDance Challenge only on YouTube #Shorts'

Question 8

like 대비 dislike의 수가 가장 적은 영상은 무엇인가? (like, dislike 값이 0인경우는 제외한다)

In [32]:
a = df[(df['likes']!=0) & (df['dislikes']!=0)].copy()
a['ratio'] = a['dislikes']/a['likes']
# 가장 적은 영상 idx
idx = a.sort_values('ratio').index[0]
df.iloc[idx].title
# [줌터뷰] *최초공개* 사부작즈🐰🐶의 비공식 이름은 아이라인즈? 꿀조합 티키타카 가득한 NCT 127 도영&정우의 줌터뷰

'[줌터뷰] *최초공개* 사부작즈🐰🐶의 비공식 이름은 아이라인즈? 꿀조합 티키타카 가득한 NCT 127 도영&정우의 줌터뷰'

Question 9

가장많은 트렌드 영상을 제작한 채널의 이름은 무엇인가? (날짜기준, 중복포함)

In [36]:
df[df.channelId ==df.channelId.value_counts().index[0]].channelTitle.unique()

# 짤툰

array(['짤툰'], dtype=object)

Question 10

20회(20일)이상 인기동영상 리스트에 포함된 동영상의 숫자는?

In [40]:
(df[['title','channelId']].value_counts()>=20).sum()

# 40

40

# 유튜브 공범컨텐츠 동영상 데이터

데이터 설명 : 유튜브 “공범” 컨텐츠 동영상 정보 ( 10분 간격 수집)

In [42]:
import pandas as pd

channel =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/channelInfo.csv')
video =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/youtube/videoInfo.csv')
display(channel.head())
display(video.head())

channelid   subcnt    viewcnt  videocnt  \
0  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  410238653       736   
1  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       736   
2  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       735   
3  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       737   
4  UCkQCwnkQfgSuPTTnw_Y7v7w  1320000  412531322       737   

                    ct channelname  
0  2021-09-30 03:01:03         꽈뚜룹  
1  2021-09-30 09:01:03         꽈뚜룹  
2  2021-09-30 15:01:03         꽈뚜룹  
3  2021-09-30 21:01:03         꽈뚜룹  
4  2021-10-01 03:01:04         꽈뚜룹

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
0  c5JQp6xafqc  1667010    30474         706            0   6587   
1  c5JQp6xafqc  1669089    30495         707            0   6589   
2  c5JQp6xafqc  1674759    30522         711            0   6596   
3  c5JQp6xafqc  1677026    30555         712            0   6604   
4  c5JQp6xafqc  1681824    30585         713            0   6600   

                    ct videoname  
0  2021-10-10 15:20:03    공범 EP1  
1  2021-10-10 15:30:03    공범 EP1  
2  2021-10-10 15:40:02    공범 EP1  
3  2021-10-10 15:50:03    공범 EP1  
4  2021-10-10 16:00:03    공범 EP1

Question 11

각 데이터의 ‘ct’컬럼을 시간으로 인식할수 있게 datatype을 변경하고 video 데이터의 videoname의 각 value 마다 몇개의 데이터씩 가지고 있는지 확인하라

In [43]:
channel['ct'] = pd.to_datetime(channel['ct'])
video['ct'] = pd.to_datetime(video['ct'])

video['videoname'].value_counts()

videoname
 공범 EP1    3492
 공범 EP2    3204
 공범 EP3    2568
 공범 EP4    2280
 공범 EP5    1562
 공범 EP6    1274
 공범 EP7     555
 공범 EP8     266
Name: count, dtype: int64

Question 12

수집된 각 video의 가장 최신화 된 날짜의 viewcount값을 출력하라

In [52]:
video.sort_values(['videoname','ct']).drop_duplicates('videoname',keep='last')[['viewcnt','videoname','ct']]

viewcnt videoname                  ct
3491   3180532    공범 EP1 2021-11-01 15:30:03
6695   2199328    공범 EP2 2021-11-01 15:30:03
9263   1671294    공범 EP3 2021-11-01 15:30:03
11543  1818493    공범 EP4 2021-11-01 15:30:03
13105  1503435    공범 EP5 2021-11-01 15:30:04
14934  1750222    공범 EP6 2021-11-01 15:30:04
13660  1630200    공범 EP7 2021-11-01 15:30:05
15200  1289088    공범 EP8 2021-11-01 15:30:05

Question 13

Channel 데이터중 2021-10-03일 이후 각 채널의 처음 기록 됐던 구독자 수(subcnt)를 출력하라

In [56]:
channel[channel.ct >= pd.to_datetime('2021-10-03')].sort_values(['channelname','ct']).drop_duplicates('channelname')[['channelname','subcnt']]

channelname   subcnt
1276    Balming Tiger    54300
1155    곽토리 kwak tori   471000
901        김농밀의 농밀한 삶     7520
12                꽈뚜룹  1330000
393            논리왕 전기   922000
1028      릴펄 Lilpearl    10100
139             야전삽짱재   257000
520   와글와글 WagleWagle    55000
647               조나단   215000
266        츄정ChuJeong   322000
774           형사!탐정되다    14900

Question 14

각채널의 2021-10-03 03:00:00 ~ 2021-11-01 15:00:00 까지 구독자수 (subcnt) 의 증가량을 구하여라

In [64]:
channel.head()

channelid   subcnt    viewcnt  videocnt                  ct  \
0  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  410238653       736 2021-09-30 03:01:03   
1  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       736 2021-09-30 09:01:03   
2  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       735 2021-09-30 15:01:03   
3  UCkQCwnkQfgSuPTTnw_Y7v7w  1310000  412531322       737 2021-09-30 21:01:03   
4  UCkQCwnkQfgSuPTTnw_Y7v7w  1320000  412531322       737 2021-10-01 03:01:04   

  channelname  
0         꽈뚜룹  
1         꽈뚜룹  
2         꽈뚜룹  
3         꽈뚜룹  
4         꽈뚜룹

In [62]:
(channel.ct.dt.strftime('%Y-%m-%d %H') == '2021-11-01 15').sum()

11

In [68]:
(channel.ct.dt.strftime('%Y-%m-%d %H') == '2021-10-03 03').sum()

11

In [75]:
start = channel[channel.ct.dt.strftime('%Y-%m-%d %H') == '2021-10-03 03']
start_df = start[['channelname','subcnt']]
start_df.columns = ['channelname','start_subcnt']

end = channel[channel.ct.dt.strftime('%Y-%m-%d %H') == '2021-11-01 15']
end_df = end[['channelname','subcnt']]
end_df.columns = ['channelname','end_subcnt']

start_end = pd.merge(start_df, end_df)
start_end['diff'] = start_end['end_subcnt'] - start_end['start_subcnt']
start_end[['channelname','diff']]

channelname   diff
0               꽈뚜룹  70000
1             야전삽짱재  11000
2        츄정ChuJeong   1000
3            논리왕 전기 -11000
4   와글와글 WagleWagle      0
5               조나단  12000
6           형사!탐정되다  10300
7        김농밀의 농밀한 삶   1540
8       릴펄 Lilpearl  11000
9     곽토리 kwak tori  -2000
10    Balming Tiger   2500

Question 15

각 비디오는 10분 간격으로 구독자수, 좋아요, 싫어요수, 댓글수가 수집된것으로 알려졌다. 공범 EP1의 비디오정보 데이터중 수집간격이 5분 이하, 20분이상인 데이터 구간( 해당 시점 전,후) 의 시각을 모두 출력하라

In [91]:
import datetime
ep1 = video[video['videoname'] == ' 공범 EP1'].sort_values('ct')

ep1[(ep1.ct.diff(1) <= datetime.timedelta(minutes=5))
    |
    (ep1.ct.diff(1) >= datetime.timedelta(minutes=20))].index


Index([721, 722, 1636], dtype='int64')

In [92]:
ep1[ep1.index.isin([720,721,722,723,1635,1636,1637])]

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
720   c5JQp6xafqc  2153676    34104         830            0   7120   
721   c5JQp6xafqc  2228250    34559         849            0   7191   
722   c5JQp6xafqc  2228250    34559         849            0   7191   
723   c5JQp6xafqc  2228669    34566         849            0   7194   
1635  c5JQp6xafqc  2706691    37893         987            0   7410   
1636  c5JQp6xafqc  2707933    37901         988            0   7410   
1637  c5JQp6xafqc  2708448    37903         988            0   7410   

                      ct videoname  
720  2021-10-12 19:20:03    공범 EP1  
721  2021-10-13 09:41:37    공범 EP1  
722  2021-10-13 09:41:37    공범 EP1  
723  2021-10-13 09:50:03    공범 EP1  
1635 2021-10-19 17:50:02    공범 EP1  
1636 2021-10-19 18:20:03    공범 EP1  
1637 2021-10-19 18:30:03    공범 EP1

Question 16

각 에피소드의 시작날짜(년-월-일)를 에피소드 이름과 묶어 데이터 프레임으로 만들고 출력하라

In [98]:
a = video.sort_values(['ct','videoname']).drop_duplicates('videoname')[['ct','videoname']]
a['ct'] = a.ct.dt.date
a.columns = [['date','videoname']]
a

date videoname
10     2021-10-07    공범 EP1
3496   2021-10-09    공범 EP2
6696   2021-10-14    공범 EP3
9264   2021-10-16    공범 EP4
11544  2021-10-21    공범 EP5
13661  2021-10-23    공범 EP6
13106  2021-10-28    공범 EP7
14935  2021-10-30    공범 EP8

Question 17

“공범” 컨텐츠의 경우 19:00시에 공개 되는것으로 알려져있다. 공개된 날의 21시의 viewcnt, ct, videoname 으로 구성된 데이터 프레임을 viewcnt를 내림차순으로 정렬하여 출력하라

In [103]:
video['time'] = video.ct.dt.hour
video[video['time'] == 21].sort_values(['videoname','ct']).drop_duplicates('videoname').sort_values('viewcnt',ascending=False)[['videoname','viewcnt','ct']]

videoname  viewcnt                  ct
14945    공범 EP8   264029 2021-10-30 21:00:08
13117    공범 EP7   252032 2021-10-28 21:00:04
9275     공범 EP4   217674 2021-10-16 21:00:04
13672    공범 EP6   213899 2021-10-23 21:00:06
11555    공범 EP5   201179 2021-10-21 21:00:04
3507     공범 EP2   148144 2021-10-09 21:00:03
6707     공범 EP3   147183 2021-10-14 21:00:04
21       공범 EP1   117340 2021-10-07 21:00:03

Question 18

video 정보의 가장 최근 데이터들에서 각 에피소드의 싫어요/좋아요 비율을 ratio 컬럼으로 만들고 videoname, ratio로 구성된 데이터 프레임을 ratio를 오름차순으로 정렬하라

In [107]:
# 가장 최근 데이터
a = video.sort_values('ct').drop_duplicates('videoname',keep='last')

a['ratio'] = a['dislikecnt']/a['likecnt']
a.sort_values('ratio')[['videoname','ratio']]

videoname     ratio
13660    공범 EP7  0.013636
14934    공범 EP6  0.015881
11543    공범 EP4  0.017971
13105    공범 EP5  0.021128
9263     공범 EP3  0.022360
3491     공범 EP1  0.027118
6695     공범 EP2  0.029844
15200    공범 EP8  0.141357

Question 19

2021-11-01 00:00:00 ~ 15:00:00까지 각 에피소드별 viewcnt의 증가량을 데이터 프레임으로 만드시오

In [110]:
start = pd.to_datetime('2021-11-01 00:00:00')
end = pd.to_datetime('2021-11-01 15:00:00')

def check(x):
    return max(x) - min(x)

a = video[(video['ct'] >= start) & (video['ct'] <= end)][['videoname','viewcnt']]
a.groupby('videoname').agg(check)

viewcnt
videoname         
 공범 EP1      13298
 공범 EP2      10300
 공범 EP3       9927
 공범 EP4       9824
 공범 EP5      10824
 공범 EP6      14141
 공범 EP7      26949
 공범 EP8      89147

Question 20

video 데이터 중에서 중복되는 데이터가 존재한다. 중복되는 각 데이터의 시간대와 videoname 을 구하여라

In [112]:
video[video.duplicated()]

videopk  viewcnt  likecnt  dislikecnt  favoritecnt  cmcnt  \
722   c5JQp6xafqc  2228250    34559         849            0   7191   
3927  23QQ1ru9YQg  1408005    20484         554            0   8567   

                      ct videoname  time  
722  2021-10-13 09:41:37    공범 EP1     9  
3927 2021-10-13 09:41:37    공범 EP2     9

# 월드컵 출전선수 골기록 데이터

데이터 설명 : 1930 ~2018년도 월드컵 출전선수 골기록

In [1]:
import pandas as pd

df= pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/worldcup/worldcupgoals.csv')
df.head()

Player  Goals                Years  Country
0  Miroslav Klose     16  2002-2006-2010-2014  Germany
1         Ronaldo     15       1998-2002-2006   Brazil
2     Gerd Muller     14            1970-1974  Germany
3   Just Fontaine     13                 1958   France
4            Pele     12  1958-1962-1966-1970   Brazil

Question 21

주어진 전체 기간의 각 나라별 골득점수 상위 5개 국가와 그 득점수를 데이터프레임형태로 출력하라

In [2]:
df.groupby('Country')['Goals'].sum().sort_values(ascending=False).to_frame().head()

Goals
Country         
Brazil       228
Germany      226
Argentina    135
Italy        123
France       115

Question 22

주어진 전체기간동안 골득점을 한 선수가 가장 많은 나라 상위 5개 국가와 그 선수 숫자를 데이터 프레임 형식으로 출력하라

In [3]:
df.groupby('Country').size().sort_values(ascending=False).to_frame().head(5)

0
Country      
Brazil     81
Germany    78
Italy      60
Argentina  59
France     58

Question 23

Years 컬럼은 년도 -년도 형식으로 구성되어있고, 각 년도는 4자리 숫자이다. 년도 표기가 4자리 숫자로 안된 케이스가 존재한다. 해당 건은 몇건인지 출력하라

In [6]:
df['yearList'] = df['Years'].str.split('-')

def check(x):
    for i in x:
        if len(i) != 4:
            return True
    return False

df['check'] = df['yearList'].apply(check)
df.check.sum()

# 45개

45

Question 24

Q3에서 발생한 예외 케이스를 제외한 데이터프레임을 df2라고 정의하고 데이터의 행의 숫자를 출력하라 (아래 문제부터는 df2로 풀이하겠습니다) 

In [16]:
# drop = True : 기존 인덱스 버리기
df2 = df[~df.check].reset_index(drop=True)
len(df2)

# 1250개

1250

In [17]:
df2

Player  Goals                Years     Country  \
0          Miroslav Klose     16  2002-2006-2010-2014     Germany   
1                 Ronaldo     15       1998-2002-2006      Brazil   
2             Gerd Muller     14            1970-1974     Germany   
3           Just Fontaine     13                 1958      France   
4                    Pele     12  1958-1962-1966-1970      Brazil   
...                   ...    ...                  ...         ...   
1245        Josip Skoblar      1                 1962  Yugoslavia   
1246          Safet Susic      1            1982-1990  Yugoslavia   
1247  Aleksandar Tirnanic      1                 1930  Yugoslavia   
1248  Djordje Vujadinovic      1                 1930  Yugoslavia   
1249         Branko Zebec      1            1954-1958  Yugoslavia   

                      yearList  check  
0     [2002, 2006, 2010, 2014]  False  
1           [1998, 2002, 2006]  False  
2                 [1970, 1974]  False  
3                       [1958]  False  
4     [1958, 1962, 1966, 1970]  False  
...                        ...    ...  
1245                    [1962]  False  
1246              [1982, 1990]  False  
1247                    [1930]  False  
1248                    [1930]  False  
1249              [1954, 1958]  False  

[1250 rows x 6 columns]

Question 25

월드컵 출전횟수를 나타내는 ‘LenCup’ 컬럼을 추가하고 4회 출전한 선수의 숫자를 구하여라

In [19]:
df2['LenCup'] = df2['yearList'].str.len()

df2['LenCup'].value_counts()[4]

# 16명

16

Question 26

Yugoslavia 국가의 월드컵 출전횟수가 2회인 선수들의 숫자를 구하여라

In [20]:
len(df2[(df2['Country'] == 'Yugoslavia') & (df2['LenCup'] == 2)])

# 7명

7

Question 27

2002년도에 출전한 전체 선수는 몇명인가?

In [30]:
df2['Years'].str.contains('2002').sum()

# 156명

156

Question 28

이름에 ‘carlos’ 단어가 들어가는 선수의 숫자는 몇 명인가? (대, 소문자 구분 x)

In [32]:
df2['Player'].str.lower().str.contains('carlos').sum()

# 13명

13

Question 29

월드컵 출전 횟수가 1회뿐인 선수들 중에서 가장 많은 득점을 올렸던 선수는 누구인가?

In [34]:
df2.head(1)

Player  Goals                Years  Country  \
0  Miroslav Klose     16  2002-2006-2010-2014  Germany   

                   yearList  check  LenCup  
0  [2002, 2006, 2010, 2014]  False       4

In [41]:
df2[(df2['LenCup'] == 1)].sort_values('Goals',ascending=False).Player.values[0]

# Just Fontaine

'Just Fontaine'

Question 30

월드컵 출전횟수가 1회 뿐인 선수들이 가장 많은 국가는 어디인가?

In [44]:
df2[df2['LenCup'] == 1].Country.value_counts().index[0]

# Brazil

'Brazil'

# 서울시 따릉이 이용정보 데이터

데이터 설명 : 서울특별시_공공자전거 시간대별 이용정보

In [102]:
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/bicycle/seoul_bi.csv')
df.head()

대여일자  대여시간  대여소번호 대여구분코드 성별 연령대코드  이용건수    운동량   탄소량     이동거리  사용시간
0  2021-06-01     0   3541    정기권  F  ~10대     1   0.00  0.00     0.00     8
1  2021-06-01     0    765    정기권  F  ~10대     1  27.21  0.35  1526.81    19
2  2021-06-01     0   2637    정기권  F  ~10대     1  41.40  0.37  1608.56    18
3  2021-06-01     0   2919    정기권  F  ~10대     1   0.00  0.00     0.00    75
4  2021-06-01     0    549    정기권  F  ~10대     1  13.04  0.17   731.55     6

Question 31

대여일자별 데이터의 수를 데이터프레임으로 출력하고, 가장 많은 데이터가 있는 날짜를 출력하라

In [103]:
a = df['대여일자'].value_counts()
display(a.sort_index().to_frame())
print(a.index[0])

count
대여일자             
2021-06-01  48081
2021-06-02  56100
2021-06-03  18985
2021-06-04  56788
2021-06-05  52730
2021-06-06  47787
2021-06-07  52623
2021-06-08  55623
2021-06-09  54507
2021-06-10  36992
2021-06-11  52089
2021-06-12  48569
2021-06-13  43131
2021-06-14  54945

2021-06-04


Question 32

각 일자의 요일을 표기하고 (‘Monday’ ~’Sunday’) ‘day_name’컬럼을 추가하고 이를 이용하여 각 요일별 이용 횟수의 총합을 데이터 프레임으로 출력하라

In [104]:
df['대여일자'] = pd.to_datetime(df['대여일자'])
df['day_name'] = df['대여일자'].dt.day_name()
df['day_name'].value_counts().to_frame()


count
day_name         
Wednesday  110607
Friday     108877
Monday     107568
Tuesday    103704
Saturday   101299
Sunday      90918
Thursday    55977

Question 33

각 요일별 가장 많이 이용한 대여소의 이용횟수와 대여소 번호를 데이터 프레임으로 출력하라

In [105]:
# 각 요일별 각 대여소 이용 횟수
a = df.groupby(['day_name','대여소번호']).size().to_frame('size').sort_values(['day_name','size'],ascending=False).reset_index()
# 가장 많이 이용한 대여소만 남기기
a.drop_duplicates('day_name',keep='first').reset_index(drop=True)

day_name  대여소번호  size
0  Wednesday    502   282
1    Tuesday    502   267
2   Thursday   2715   137
3     Sunday    502   372
4   Saturday    502   378
5     Monday    502   242
6     Friday    502   277

Question 34

나이대별 대여구분 코드의 (일일권/전체횟수) 비율을 구한 후 가장 높은 비율을 가지는 나이대를 확인하라. 일일권의 경우 일일권 과 일일권(비회원)을 모두 포함하라

In [106]:
df['대여구분코드'].unique()

array(['정기권', '일일권', '단체권', '일일권(비회원)'], dtype=object)

In [107]:
# 나이대별 일일권 구매 횟수
daily = df[df['대여구분코드'].isin(['일일권','일일권(비회원)'])].연령대코드.value_counts().sort_index()

# 나이대별 사람 수
total = df.연령대코드.value_counts().sort_index()

# 비율
ratio = (daily/total).sort_values(ascending=False)
ratio.index[0]

# ~10대

'~10대'

Question 35

연령대별 평균 이동거리를 구하여라

In [109]:
df[['연령대코드','이동거리']].groupby('연령대코드').mean()

이동거리
연령대코드             
20대    3211.890552
30대    3341.443859
40대    3514.857416
50대    3593.668100
60대    3538.145737
70대~   3085.039641
~10대   2634.426279

Question 36

연령대 코드가 20대인 데이터를 추출하고,이동거리값이 추출한 데이터의 이동거리값의 평균 이상인 데이터를 추출한다.최종 추출된 데이터를 대여일자, 대여소 번호 순서로 내림차순 정렬 후 1행부터 200행까지의 탄소량의 평균을 소숫점 3째 자리까지 구하여라

In [110]:
df['탄소량'].unique()

array(['0.00', '0.35', '0.37', ..., '12.39', '8.96', '17.01'],
      dtype=object)

In [113]:
df[df['탄소량'] == '\\N']

대여일자  대여시간  대여소번호 대여구분코드 성별 연령대코드  이용건수 운동량 탄소량  이동거리  사용시간  \
17     2021-06-01     0   1167    정기권  F   20대     1  \N  \N   0.0    20   
313    2021-06-01     0   2396    정기권  M   30대     1  \N  \N   0.0    15   
416    2021-06-01     0    250    정기권  M   40대     1  \N  \N   0.0    56   
694    2021-06-01     0   2356    일일권  M   30대     1  \N  \N   0.0    22   
897    2021-06-01     0    125    정기권  M   40대     1  \N  \N   0.0    69   
...           ...   ...    ...    ... ..   ...   ...  ..  ..   ...   ...   
676885 2021-06-14    22   1282    정기권  F   50대     1  \N  \N   0.0   117   
677276 2021-06-14    22   2548    정기권  M   40대     1  \N  \N   0.0     9   
677732 2021-06-14    23   2377    정기권  M   20대     1  \N  \N   0.0    30   
678029 2021-06-14    23   4206    정기권  M   30대     1  \N  \N   0.0    55   
678137 2021-06-14    23   1958    정기권  M   50대     1  \N  \N   0.0    59   

       day_name  
17      Tuesday  
313     Tuesday  
416     Tuesday  
694     Tuesday  
897     Tuesday  
...         ...  
676885   Monday  
677276   Monday  
677732   Monday  
678029   Monday  
678137   Monday  

[3875 rows x 12 columns]

In [128]:
# 20대
twenty = df[df['연령대코드'] == '20대']
twenty
# 이동거리값이 평균 이상
mean_up = twenty[twenty['이동거리'] >= twenty['이동거리'].mean()]
# 대여일자, 대여소 번호 순서로 내림차순 정렬 후 200행의 탄소량
carbon = mean_up.sort_values(['대여일자','대여소번호'],ascending=False)[:200]['탄소량']
# 평균
carbon = carbon.astype('float')
round(carbon.mean(),3)

# 1.613

1.613

Question 37

6월 7일 ~10대의 “이용건수”의 중앙값은?

In [130]:
df['대여일자']

0        2021-06-01
1        2021-06-01
2        2021-06-01
3        2021-06-01
4        2021-06-01
            ...    
678945   2021-06-14
678946   2021-06-14
678947   2021-06-14
678948   2021-06-14
678949   2021-06-14
Name: 대여일자, Length: 678950, dtype: datetime64[ns]

In [133]:
df[(df['대여일자'] == pd.to_datetime('2021-06-07'))&(df['연령대코드'] == '~10대')].이용건수.median()

# 1.0

1.0

Question 38

평일 (월~금) 출근 시간대(오전 6,7,8시)의 대여소별 이용 횟수를 구해서 데이터 프레임 형태로 표현한 후 각 대여시간별 이용 횟수의 상위 3개 대여소와 이용횟수를 출력하라

In [165]:
a = df[(df['대여일자'].dt.day_name().isin(['Monday','Tuesday','Wednesday','Thursday','Friday'])) & (df['대여시간'].isin([6,7,8]))]
a = a.groupby(['대여시간','대여소번호']).size().to_frame('이용 횟수')
display(a)
a.sort_values(['대여시간','이용 횟수'],ascending=False).groupby('대여시간').head(3)

이용 횟수
대여시간 대여소번호       
6    102        5
     103        8
     106        1
     109        4
     111        5
...           ...
8    4864       4
     4865      12
     4867       3
     4868       3
     9980       1

[6236 rows x 1 columns]

이용 횟수
대여시간 대여소번호       
8    2701     119
     646      115
     1152      92
7    259      104
     230       77
     726       77
6    2744      45
     1125      40
     1028      36

Question 39

이동거리의 평균 이상의 이동거리 값을 가지는 데이터를 추출하여 추출데이터의 이동거리의 표본표준편차 값을 구하여라

In [166]:
df[df['이동거리']>=df['이동거리'].mean()]['이동거리'].std()

# 5092.139707505356

5092.139707505356

Question 40

남성(‘M’ or ‘m’)과 여성(‘F’ or ‘f’)의 이동거리값의 평균값을 구하여라

In [172]:
male = df[df['성별'].isin(['M','m'])]['이동거리'].mean()
print(male)

# 3209.1108714530546

female = df[df['성별'].isin(['F','f'])]['이동거리'].mean()
print(female)

# 3468.5750252981957

3209.1108714530546
3468.5750252981957


# 전세계 행복도 지표 데이터

데이터 설명 : 전세계 행복도 지표 조사

In [174]:
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/happy2/happiness.csv',encoding='utf-8')
df.head()

행복랭킹          나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도    관대함  부패에 대한인식  \
0     1      Finland  7.769  1.340  1.587  0.986    0.596  0.153     0.393   
1     2      Denmark  7.600  1.383  1.573  0.996    0.592  0.252     0.410   
2     3       Norway  7.554  1.488  1.582  1.028    0.603  0.271     0.341   
3     4      Iceland  7.494  1.380  1.624  1.026    0.591  0.354     0.118   
4     5  Netherlands  7.488  1.396  1.522  0.999    0.557  0.322     0.298   

     년도  
0  2019  
1  2019  
2  2019  
3  2019  
4  2019

Question 41

데이터는 2018년도와 2019년도의 전세계 행복 지수를 표현한다. 각년도의 행복랭킹 10위를 차지한 나라의 행복점수의 평균을 구하여라

In [177]:
df[df['행복랭킹'] == 10].점수.mean()

# 7.259

7.259

Question 42

데이터는 2018년도와 2019년도의 전세계 행복 지수를 표현한다. 각년도의 행복랭킹 50위이내의 나라들의 각각의 행복점수 평균을 데이터프레임으로 표시하라

In [179]:
df[df['행복랭킹'] <= 50][['년도','점수']].groupby('년도').mean()

점수
년도           
2018  6.64678
2019  6.67002

Question 43

2018년도 데이터들만 추출하여 행복점수와 부패에 대한 인식에 대한 상관계수를 구하여라

In [180]:
df.head()

행복랭킹          나라명     점수  상대GDP  사회적지원  행복기대치  선택의 자유도    관대함  부패에 대한인식  \
0     1      Finland  7.769  1.340  1.587  0.986    0.596  0.153     0.393   
1     2      Denmark  7.600  1.383  1.573  0.996    0.592  0.252     0.410   
2     3       Norway  7.554  1.488  1.582  1.028    0.603  0.271     0.341   
3     4      Iceland  7.494  1.380  1.624  1.026    0.591  0.354     0.118   
4     5  Netherlands  7.488  1.396  1.522  0.999    0.557  0.322     0.298   

     년도  
0  2019  
1  2019  
2  2019  
3  2019  
4  2019

In [185]:
df[df['년도'] == 2018][['점수','부패에 대한인식']].corr().iloc[0,1]

# 0.40529152271510027

0.40529152271510027

Question 44

2018년도와 2019년도의 행복랭킹이 변화하지 않은 나라명의 수를 구하여라

In [190]:
len(df[['행복랭킹','나라명']]) - len(df[['행복랭킹','나라명']].drop_duplicates())

# 15

15

Question 45

2019년도 데이터들만 추출하여 각변수간 상관계수를 구하고 내림차순으로 정렬한 후 상위 5개를 데이터 프레임으로 출력하라. 컬럼명은 v1,v2,corr으로 표시하라

In [227]:
a = df[df['년도'] == 2019].drop('나라명',axis=1).corr().stack().to_frame()
a = a[a != 1].sort_values(0,ascending=False).drop_duplicates(0).head(5).reset_index()
a.columns = ['v1','v2','corr']
a

v1     v2      corr
0  행복기대치  상대GDP  0.835462
1  상대GDP     점수  0.793883
2  행복기대치     점수  0.779883
3  사회적지원     점수  0.777058
4  사회적지원  상대GDP  0.754906

Question 46

각 년도별 하위 행복점수의 하위 5개 국가의 평균 행복점수를 구하여라

In [196]:
df.groupby('년도').tail(5).groupby('년도')['점수'].mean()

년도
2018    3.1800
2019    3.1408
Name: 점수, dtype: float64

Question 47

2019년 데이터를 추출하고 해당데이터의 상대 GDP 평균 이상의 나라들과 평균 이하의 나라들의 행복점수 평균을 각각 구하고 그 차이값을 출력하라

In [240]:
a = a.loc[df['년도'] == 2019]
over = a[a.상대GDP >= a.상대GDP.mean()]['점수'].mean()
under = a[a.상대GDP <= a.상대GDP.mean()]['점수'].mean()

over - under

# 1.390483957219251

1.390483957219251

Question 48

각년도의 부패에 대한인식을 내림차순 정렬했을때 상위 20개 국가의 부패에 대한인식의 평균을 구하여라

In [245]:
df.sort_values(['년도','부패에 대한인식'],ascending=False).groupby('년도').head(20).groupby('년도')['부패에 대한인식'].mean()

년도
2018    0.3267
2019    0.3201
Name: 부패에 대한인식, dtype: float64

Question 49

2018년도 행복랭킹 50위 이내에 포함됐다가 2019년 50위 밖으로 밀려난 국가의 숫자를 구하여라

In [264]:
a = set(df[(df['년도'] == 2018) & (df['행복랭킹'] <= 50)].나라명)
b = set(df[(df['년도'] == 2019) & (df['행복랭킹'] <= 50)].나라명)
len(a-b)

# 4

4

Question 50

2018년,2019년 모두 기록이 있는 나라들 중 년도별 행복점수가 가장 증가한 나라와 그 증가 수치는?

In [279]:
country = df.나라명.value_counts()
country = country[country>=2].index
country

a = df[df.나라명.isin(country)].reset_index(drop=True)
# 2018년 점수 음수로
a.loc[a.년도 == 2018,'점수'] = a.loc[a.년도 == 2018,'점수']*-1
a = a.groupby('나라명')['점수'].sum().sort_values().to_frame()
print(a.index[-1],a.values[-1])

# Burundi 0.87

Burundi [0.87]


# 지역구 에너지 소비량 데이터

데이터 설명 : 기온, 습도, 바람풍속에 따른 도시의 3개 지역구의 에너지 소비량

In [1]:
import pandas as pd
df= pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/consum/Tetuan%20City%20power%20consumption.csv')
df.head()

DateTime  Temperature  Humidity  Wind Speed  general diffuse flows  \
0  1/1/2017 0:00        6.559      73.8       0.083                  0.051   
1  1/1/2017 0:10        6.414      74.5       0.083                  0.070   
2  1/1/2017 0:20        6.313      74.5       0.080                  0.062   
3  1/1/2017 0:30        6.121      75.0       0.083                  0.091   
4  1/1/2017 0:40        5.921      75.7       0.081                  0.048   

   diffuse flows  Zone 1 Power Consumption  Zone 2  Power Consumption  \
0          0.119               34055.69620                16128.87538   
1          0.085               29814.68354                19375.07599   
2          0.100               29128.10127                19006.68693   
3          0.096               28228.86076                18361.09422   
4          0.085               27335.69620                17872.34043   

   Zone 3  Power Consumption  
0                20240.96386  
1                20131.08434  
2                19668.43373  
3                18899.27711  
4                18442.40964

Question 51

DateTime컬럼을 통해 각 월별로 몇개의 데이터가 있는지 데이터 프레임으로 구하여라

In [2]:
df['DateTime'] = pd.to_datetime(df['DateTime'])
df['DateTime'].dt.month.value_counts().sort_index().to_frame()

count
DateTime       
1          4464
2          4032
3          4464
4          4320
5          4464
6          4320
7          4464
8          4464
9          4320
10         4464
11         4320
12         4320

Question 52

3월달의 각 시간대별 온도의 평균들 중 가장 낮은 시간대의 온도를 출력하라

In [3]:
a = df[df['DateTime'].dt.month == 3]
a.groupby(df.DateTime.dt.hour)['Temperature'].mean().sort_values().iloc[0]

# 11.506612903225806

11.506612903225806

Question 53

3월달의 각 시간대별 온도의 평균들 중 가장 높은 시간대의 온도를 출력하라

In [10]:
df[df['DateTime'].dt.month == 3].groupby(df['DateTime'].dt.hour).Temperature.mean().sort_values().iloc[-1]

# 18.393602150537635

18.393602150537635

Question 54

Zone 1 Power Consumption 컬럼의 value값의 크기가 Zone 2 Power Consumption 컬럼의 value값의 크기보다 큰 데이터들의 Humidity의 평균을 구하여라

In [11]:
df[df['Zone 1 Power Consumption'] > df['Zone 2  Power Consumption']].Humidity.mean()

# 68.23624448055094

68.23624448055052

Question 55

각 zone의 에너지 소비량의 상관관계를 구해서 데이터 프레임으로 표기하라

In [12]:
df.columns

Index(['DateTime', 'Temperature', 'Humidity', 'Wind Speed',
       'general diffuse flows', 'diffuse flows', 'Zone 1 Power Consumption',
       'Zone 2  Power Consumption', 'Zone 3  Power Consumption'],
      dtype='object')

In [13]:
df[['Zone 1 Power Consumption', 'Zone 2  Power Consumption', 'Zone 3  Power Consumption']].corr()

Zone 1 Power Consumption  \
Zone 1 Power Consumption                   1.000000   
Zone 2  Power Consumption                  0.834519   
Zone 3  Power Consumption                  0.750733   

                           Zone 2  Power Consumption  \
Zone 1 Power Consumption                    0.834519   
Zone 2  Power Consumption                   1.000000   
Zone 3  Power Consumption                   0.570932   

                           Zone 3  Power Consumption  
Zone 1 Power Consumption                    0.750733  
Zone 2  Power Consumption                   0.570932  
Zone 3  Power Consumption                   1.000000

Question 56

Temperature의 값이 10미만의 경우 A, 10이상 20미만의 경우 B,20이상 30미만의 경우 C, 그 외의 경우 D라고 할때 각 단계의 데이터 숫자를 구하여라

In [14]:
def Tem(x):
    if x < 10:
        return 'A'
    elif x < 20:
        return 'B'
    elif x < 30:
        return 'C'
    else:
        return 'D'

df['Temperature'].apply(Tem).value_counts()

Temperature
B    26993
C    21105
A     2874
D     1444
Name: count, dtype: int64

Question 57

6월 데이터중 12시의 Temperature의 표준편차를 구하여라

In [15]:
df[(df['DateTime'].dt.month == 6) & (df['DateTime'].dt.hour == 12)].Temperature.std()

# 2.0499417827951025

2.0499417827951025

Question 58

6월 데이터중 12시의 Temperature의 분산을 구하여라

In [16]:
df[(df['DateTime'].dt.month == 6) & (df['DateTime'].dt.hour == 12)].Temperature.var()

# 4.202261312849163

4.202261312849163

Question 59

Temperature의 평균이상의 Temperature의 값을 가지는 데이터를 Temperature를 기준으로 정렬 했을때 4번째 행의 Humidity 값은?

In [20]:
df[df['Temperature'] >= df['Temperature'].mean()].sort_values('Temperature').Humidity.values[3]

# 87.9

87.9

Question 60

Temperature의 중간값 이상의 Temperature의 값을 가지는 데이터를Temperature를 기준으로 정렬 했을때 4번째 행의 Humidity 값은?

In [21]:
df[df['Temperature'] >= df['Temperature'].median()].sort_values('Temperature').Humidity.values[3]

# 80.3

80.3

# 포켓몬 정보 데이터

데이터 설명 : 포켓몬 정보

In [22]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/pok/Pokemon.csv')
df.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Question 61

Legendary 컬럼은 전설포켓몬 유무를 나타낸다.전설포켓몬과 그렇지 않은 포켓몬들의 HP평균의 차이를 구하여라

In [25]:
yes = df[df['Legendary']].HP.mean()
no = df[~df['Legendary']].HP.mean()
yes - no

# 25.55614861329147

25.55614861329147

Question 62

Type 1은 주속성 Type 2 는 부속성을 나타낸다. 가장 많은 부속성 종류는 무엇인가?

In [30]:
df['Type 2'].value_counts().index[0]

# Flying

'Flying'

Question 63

가장 많은 Type 1 의 종의 평균 Attack 을 평균 Defense로 나눈값은?

In [31]:
a = df['Type 1'].value_counts().index[0]
df[df['Type 1'] == a].Attack.mean()/df[df['Type 1'] == a].Defense.mean()

# a = df['Type 1'].value_counts().index[0]
df[df['Type 1'] == a].Attack.mean()/df[df['Type 1'] == a].Defense.mean()

# 

1.0165238678090576

Question 64

포켓몬 세대(Generation) 중 가장많은 Legendary를 보유한 세대는 몇세대인가?

In [39]:
df[['Generation','Legendary']].groupby('Generation').sum().sort_values('Legendary').index[-1]

# df[df.Legendary==True].Generation.value_counts().index[0]
# 3

3

Question 65

‘HP’, ‘Attack’, ‘Defense’, ‘Sp. Atk’, ‘Sp. Def’, ‘Speed’ 간의 상관 계수중 가장 절댓값이 큰 두 변수와 그 값을 구하여라

In [52]:
df[['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']].corr().stack().reset_index().sort_values(0,ascending=False).iloc[6]

level_0     Defense
level_1     Sp. Def
0          0.510747
Name: 16, dtype: object

Question 66

각 Generation의 Attack으로 오름차순 정렬시 상위 3개 데이터들(18개)의 Attack의 전체 평균을 구하여라

In [55]:
df.sort_values(['Generation','Attack']).groupby('Generation').head(3).Attack.mean()

# 19.5

19.5

Question 67

각 Generation의 Attack으로 내림차순 정렬시 상위 5개 데이터들(30개)의 Attack의 전체 평균을 구하여라

In [58]:
df.sort_values(['Generation','Attack'],ascending=False).groupby('Generation').head(5).Attack.mean()

# 157.23333333333332

157.23333333333332

Question 68

가장 흔하게 발견되는 (Type1 , Type2) 의 쌍은 무엇인가?

In [64]:
df[['Type 1','Type 2']].value_counts().head(1)

Type 1  Type 2
Normal  Flying    24
Name: count, dtype: int64

Question 69

한번씩만 존재하는 (Type1 , Type2)의 쌍의 갯수는 몇개인가?

In [72]:
a = df[['Type 1','Type 2']].value_counts()
len(a[a == 1])

# 39

39

Question 70

한번씩만 존재하는 (Type1 , Type2)의 쌍을 각 세대(Generation)은 각각 몇개씩 가지고 있는가?

In [89]:
a = df[['Type 1','Type 2']].value_counts()
a_ls = a[a == 1]

lst = []
for value in a_ls.index:
    t1 = value[0]
    t2 = value[1]
    new = df[(df['Type 1'] == t1) & (df['Type 2'] == t2)]
    lst.append(new)

pd.concat(lst).reset_index(drop=True).Generation.value_counts().sort_index()

Generation
1     1
2     4
3     5
4    13
5     7
6     9
Name: count, dtype: int64

# 대한민국 체력장 데이터

데이터 설명 : 대한민국 국민 체력장 평가

In [92]:
import pandas as pd
df =pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/body/body.csv')
df.head()

측정나이 측정회원성별  신장 : cm  체중 : kg  체지방율 : %  이완기혈압(최저) : mmHg  \
0  59.0      M    175.0     70.6      19.2              91.0   
1  40.0      F    161.4     45.8      21.7              62.0   
2  27.0      M    176.6     84.8      19.2              87.0   
3  38.0      M    167.6     76.7      24.7              63.0   
4  21.0      M    165.2     66.2      21.5              83.0   

   수축기혈압(최고) : mmHg  악력D : kg  앉아윗몸앞으로굽히기 : cm  교차윗몸일으키기 : 회  제자리 멀리뛰기 : cm 등급  
0             150.0      40.6             12.2          30.0          179.0  C  
1             119.0      22.1             16.0          32.0          165.0  B  
2             153.0      45.9             13.2          61.0          216.0  B  
3             132.0      43.6             16.0          45.0          231.0  A  
4             106.0      33.5             10.6          46.0          198.0  C

Question 71

전체데이터의 수축기혈압(최고) - 이완기혈압(최저)의 평균을 구하여라

In [96]:
(df['수축기혈압(최고) : mmHg'] - df['이완기혈압(최저) : mmHg']).mean()

# 52.19539414750672

52.19539414750672

Question 72

50~59세의 신장평균을 구하여라

In [98]:
df[(df['측정나이'] >= 50) & (df['측정나이'] <= 59)]['신장 : cm'].mean()

# 164.07490107405314

164.07490107405314

Question 73

연령대 (20~29 : 20대 …) 별 인원수를 구하여라

In [101]:
(df['측정나이']//10*10).value_counts()


측정나이
20.0    5831
30.0    2660
40.0    1801
50.0    1769
60.0    1335
Name: count, dtype: int64

Question 74

연령대 (20~29 : 20대 …) 별 등급의 숫자를 데이터 프레임으로 표현하라

In [106]:
df['age'] = df['측정나이']//10*10
df[['age','등급']].value_counts().sort_index().to_frame()

# df.groupby(['연령대','등급'],as_index=False).size()

count
age  등급       
20.0 A    1585
     B    1443
     C    1455
     D    1348
30.0 A     743
     B     697
     C     626
     D     594
40.0 A     386
     B     428
     C     455
     D     532
50.0 A     321
     B     410
     C     474
     D     564
60.0 A     314
     B     371
     C     339
     D     311

Question 75

남성 중 A등급과 D등급의 체지방률 평균의 차이(큰 값에서 작은 값의 차)를 구하여라

In [109]:
man = df[df['측정회원성별'] == 'M']
abs(man[man['등급'] == 'A']['체지방율 : %'].mean() - man[man['등급'] == 'D']['체지방율 : %'].mean())

# 7.932086486137521

7.932086486137521

Question 76

여성 중 A등급과 D등급의 체중의 평균의 차이(큰 값에서 작은 값의 차)를 구하여라

In [115]:
woman = df[df['측정회원성별'] == 'F']
abs(woman[woman['등급'] == 'A']['체중 : kg'].mean() - woman[woman['등급'] == 'D']['체중 : kg'].mean())

# 5.176211590296504

5.176211590296504

Question 77

bmi는 자신의 몸무게(kg)를 키의 제곱(m)으로 나눈값이다. 데이터의 bmi 를 구한 새로운 컬럼을 만들고 남성의 bmi 평균을 구하여라

In [121]:
df['bmi'] = df['체중 : kg']/(df['신장 : cm']/100)**2
df[df['측정회원성별'] == 'M'].bmi.mean()

# 24.46134409819308

24.46134409819308

Question 78

bmi보다 체지방율이 높은 사람들의 체중평균을 구하여라

In [123]:
df[df['체지방율 : %'] > df['bmi']]['체중 : kg'].mean()

# 61.7408806392542

61.7408806392542

Question 79

남성과 여성의 악력 평균의 차이를 구하여라

In [124]:
man = df[df['측정회원성별'] == 'M']['악력D : kg'].mean()
woman = df[df['측정회원성별'] == 'F']['악력D : kg'].mean()
man - woman

# 17.55954185047464

17.55954185047464

Question 80

남성과 여성의 교차윗몸일으키기 횟수의 평균의 차이를 구하여라

In [125]:
man = df[df['측정회원성별'] == 'M']['교차윗몸일으키기 : 회'].mean()
woman = df[df['측정회원성별'] == 'F']['교차윗몸일으키기 : 회'].mean()
man - woman

# 14.243156833157627

14.243156833157627

In [122]:
df.head(1)

측정나이 측정회원성별  신장 : cm  체중 : kg  체지방율 : %  이완기혈압(최저) : mmHg  \
0  59.0      M    175.0     70.6      19.2              91.0   

   수축기혈압(최고) : mmHg  악력D : kg  앉아윗몸앞으로굽히기 : cm  교차윗몸일으키기 : 회  제자리 멀리뛰기 : cm  \
0             150.0      40.6             12.2          30.0          179.0   

  등급   age        bmi  
0  C  50.0  23.053061

# 기온 강수량 데이터

데이터 설명 : 2020년도 이화동(서울) , 수영동(부산)의 시간단위의 기온과 강수량

In [126]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/weather/weather2.csv")
df.head()

time  이화동강수  이화동기온  수영동강수  수영동기온
0  2020-01-01 00:00:00    0.1   -3.9    0.0    2.5
1  2020-01-01 01:00:00    0.0   -3.1    0.0    3.4
2  2020-01-01 02:00:00    0.0   -1.8    0.0    3.8
3  2020-01-01 03:00:00    0.0   -0.8    0.0    4.6
4  2020-01-01 04:00:00    0.0   -0.1    0.0    5.1

Question 81

여름철(6월,7월,8월) 이화동이 수영동보다 높은 기온을 가진 시간대는 몇개인가?

In [131]:
df['time'] = pd.to_datetime(df['time'])
a = df[df['time'].dt.month.isin([6,7,8])]
len(a[a['이화동기온'] > a['수영동기온']]['time'])

# 1415

1415

Question 82

이화동과 수영동의 최대강수량의 시간대를 각각 구하여라

In [134]:
ehwa = df[df['이화동강수'] == df['이화동강수'].max()]['time']
print(ehwa)

soyoung = df[df['수영동강수'] == df['수영동강수'].max()]['time']
print(soyoung)

6561   2020-09-30 09:00:00
Name: time, dtype: datetime64[ns]
4908   2020-07-23 12:00:00
Name: time, dtype: datetime64[ns]


# 서비스 이탈예측 데이터

데이터 설명 : 고객의 신상정보 데이터를 통한 회사 서비스 이탈 예측 (종속변수 : Exited)

In [135]:
import pandas as pd
#데이터 로드
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/churn/train.csv")
df.head()

RowNumber  CustomerId     Surname  CreditScore Geography  Gender  Age  \
0       6842    15793491  Cherkasova          714   Germany    Male   26   
1       8963    15607874       Keane          687    France    Male   38   
2       7047    15737627      Rivero          589   Germany  Female   20   
3       7503    15697844  Whitehouse          721     Spain  Female   32   
4       3439    15722404   Carpenter          445    France  Female   30   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       3  119545.48              2          1               0   
1       0  144450.58              1          0               1   
2       2  121093.29              2          1               0   
3      10       0.00              1          1               0   
4       3       0.00              2          1               1   

   EstimatedSalary  Exited  
0         65482.94       0  
1        137276.83       0  
2          3529.72       0  
3        136119.96       1  
4        127939.19       0

Question 83

남성 이탈(Exited)이 가장 많은 국가(Geography)는 어디이고 이탈 인원은 몇명인가?

In [140]:
df[(df['Gender'] == 'Male')&(df['Exited'] == 1)].Geography.value_counts().head(1)

# Germany 287

Geography
Germany    287
Name: count, dtype: int64

Question 84

카드를 소유(HasCrCard ==1)하고 있으면서 활성멤버(IsActiveMember ==1) 인 고객들의 평균 나이를 소숫점이하 4자리까지 구하여라?

In [142]:
round(df[(df['HasCrCard'] == 1) & (df['IsActiveMember'] == 1)].Age.mean(),4)

# 39.6102

39.6102

Question 85

Balance 값이 중간값 이상을 가지는 고객들의 CreditScore의 표준편차를 소숫점이하 3자리까지 구하여라

In [146]:
round(df[df['Balance'] >= df['Balance'].median()]['CreditScore'].std(),3)

# 97.295

97.295

# 성인 건강검진 데이터

데이터 설명 : 2018년도 성인의 건강검 진데이터 (흡연상태 1- 흡연, 0-비흡연 )

In [147]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/smoke/train.csv")
df.head()

성별코드  연령대코드(5세단위)  신장(5Cm단위)  체중(5Kg단위)  허리둘레  시력(좌)  시력(우)  청력(좌)  청력(우)  \
0    F           55        145         55  73.0    0.7    0.5    1.0    1.0   
1    M           40        180         55  74.0    1.5    1.5    1.0    1.0   
2    F           55        150         50  72.0    1.0    0.2    1.0    1.0   
3    M           40        170         85  89.0    1.0    0.8    1.0    1.0   
4    F           40        155         45  62.0    0.5    1.0    1.0    1.0   

   수축기혈압  ...   혈색소  요단백  혈청크레아티닌  (혈청지오티)AST  (혈청지오티)ALT  감마지티피  흡연상태  \
0  129.0  ...  12.9  1.0      0.9       172.0       209.0   15.0     0   
1  102.0  ...  15.6  1.0      0.8        30.0        19.0   23.0     1   
2  116.0  ...  14.1  1.0      1.0        31.0        19.0   10.0     0   
3  124.0  ...  16.7  1.0      1.0        26.0        38.0  108.0     1   
4  127.0  ...  12.9  1.0      0.7        20.0        11.0   13.0     0   

   구강검진수검여부  치아우식증유무  치석  
0         Y      0.0   Y  
1         Y      0.0   Y  
2         Y      0.0   N  
3         Y      1.0   Y  
4         Y      0.0   N  

[5 rows x 26 columns]

Question 86

수축기혈압과 이완기 혈압기 수치의 차이를 새로운 컬럼(‘혈압차’) 으로 생성하고, 연령대 코드별 각 그룹 중 ‘혈압차’ 의 분산이 5번째로 큰 연령대 코드를 구하여라

In [150]:
df['혈압차'] = df['수축기혈압'] - df['이완기혈압']
df.groupby('연령대코드(5세단위)')['혈압차'].var().sort_values().index[-5]

# 60

60

Question 87

비만도를 나타내는 지표인 WHtR는 허리둘레 / 키로 표현한다. 일반적으로 0.58이상이면 비만으로 분류한다. 데이터중 WHtR 지표상 비만인 인원의 남/여 비율을 구하여라

In [153]:
df['WHtR'] = df['허리둘레']/df['신장(5Cm단위)']
man = len(df[(df['WHtR'] >= 0.58) & (df['성별코드'] == 'M')])
woman = len(df[(df['WHtR'] >= 0.58) & (df['성별코드'] == 'F')])
man/woman

# 1.1693877551020408

1.1693877551020408

# 자동차 보험가입 예측데이터

데이터 설명 : 자동차 보험 가입 예측

In [154]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/insurance/train.csv")
df.head()

id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0  188957  Female   25                1         17.0                   1   
1  275631    Male   20                1         22.0                   0   
2  329036  Female   40                1         41.0                   0   
3  227288  Female   28                1         30.0                   1   
4  300441  Female   50                1         28.0                   0   

  Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  Vintage  \
0    < 1 Year             No         38785.0                 152.0       59   
1    < 1 Year             No          2630.0                 160.0      113   
2    1-2 Year            Yes         33165.0                 124.0      188   
3    < 1 Year             No         29038.0                 152.0      233   
4    1-2 Year            Yes         31325.0                 124.0      258   

   Response  
0         0  
1         0  
2         0  
3         0  
4         0

Question 88

Vehicle_Age 값이 2년 초과인 사람들만 필터링 하고 그중에서
Annual_Premium 값이 전체 데이터의 중간값 이상인 사람들을 찾고, 그들의 Vintage값의 평균을 구하여라

In [156]:
df[(df['Vehicle_Age'] == '> 2 Years') & (df['Annual_Premium'] >= df['Annual_Premium'].median())].Vintage.mean()

# 154.43647182359118

154.43647182359118

Question 89

vehicle_age에 따른 각 성별(gender)그룹의 Annual_Premium값의 평균을 구하여 아래 테이블과 동일하게 구현하라

In [164]:
df.head(1)

id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0  188957  Female   25                1         17.0                   1   

  Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  Vintage  \
0    < 1 Year             No         38785.0                 152.0       59   

   Response  
0         0

In [172]:
a = df.groupby(['Vehicle_Age','Gender'],as_index=False)['Annual_Premium'].mean()
a.pivot_table(index='Vehicle_Age',columns='Gender',values='Annual_Premium')

Gender             Female          Male
Vehicle_Age                            
1-2 Year     30762.245001  30413.088469
< 1 Year     29972.286702  30310.982212
> 2 Years    36108.366374  35303.870627

# 핸드폰 가격 예측데이터

데이터 설명 : 핸드폰 가격예측 (price_range컬럼 0(저렴) ~3(매우비쌈) 범위 )

In [173]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/mobile/train.csv")
df.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

Question 90

price_range 의 각 value를 그룹핑하여 각 그룹의 n_cores 의 빈도가 가장높은 value와 그 빈도수를 구하여라

In [182]:
df[['price_range','n_cores']].groupby(['price_range','n_cores']).size().sort_values().groupby('price_range').tail(1)

price_range  n_cores
0            2          69
3            5          70
2            4          73
1            1          76
dtype: int64

Question 91

price_range 값이 3인 그룹에서 상관관계가 2번째로 높은 두 컬럼과 그 상관계수를 구하여라

In [194]:
a = df[df['price_range'] == 3].corr().stack()
a[a != 1].sort_values().reset_index().iloc[-2]

level_0          pc
level_1          fc
0          0.635166
Name: 378, dtype: object

# 비행탑승 경험 만족도 데이터

데이터 설명 : 비행탑승 경험 만족도 (satisfaction 컬럼 : ‘neutral or dissatisfied’ or satisfied ) (83123, 24) shape

In [195]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/airline/train.csv")
df.head()

id  Gender   Customer Type  Age   Type of Travel     Class  \
0   55882    Male  Loyal Customer   42  Business travel  Business   
1   31906  Female  Loyal Customer   31  Business travel  Business   
2   85481    Male  Loyal Customer   51  Personal Travel       Eco   
3  102134    Male  Loyal Customer   44  Business travel  Business   
4   31580    Male  Loyal Customer   33  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0              473                      1                                  1   
1             3271                      4                                  4   
2              214                      5                                  0   
3             3097                      1                                  1   
4             3286                      1                                  1   

   Ease of Online booking  ...  Inflight entertainment  On-board service  \
0                       3  ...                       5                 5   
1                       4  ...                       3                 3   
2                       5  ...                       3                 4   
3                       3  ...                       4                 4   
4                       1  ...                       4                 2   

   Leg room service  Baggage handling  Checkin service  Inflight service  \
0                 5                 5                3                 5   
1                 5                 5                3                 4   
2                 4                 4                4                 4   
3                 4                 4                3                 4   
4                 5                 2                5                 1   

   Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  \
0            3                           3                       0.0   
1            3                           1                       5.0   
2            3                           2                       0.0   
3            5                           0                       0.0   
4            4                           0                       0.0   

   satisfaction  
0     satisfied  
1     satisfied  
2     satisfied  
3     satisfied  
4     satisfied  

[5 rows x 24 columns]

Question 92

Arrival Delay in Minutes 컬럼이 결측치인 데이터들 중 ‘neutral or dissatisfied’ 보다 ‘satisfied’의 수가 더 높은 Class는 어디 인가?

In [198]:
df.satisfaction.unique()

array(['satisfied', 'neutral or dissatisfied'], dtype=object)

In [202]:
df.Class.unique()

array(['Business', 'Eco', 'Eco Plus'], dtype=object)

In [205]:
a = df[df['Arrival Delay in Minutes'].isnull()].groupby(['Class','satisfaction']).size().unstack()
a[a['neutral or dissatisfied'] < a['satisfied']]

# Business	

satisfaction  neutral or dissatisfied  satisfied
Class                                           
Business                           36         76

# 수질 음용성 여부 데이터

데이터 설명 : 수질 음용성 여부 (Potablillity 컬럼 : 0 ,1 )

In [206]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/waters/train.csv")
df.head()

ph    Hardness        Solids  Chloramines     Sulfate  Conductivity  \
0  7.918150  214.186611  23823.492888     7.290878  341.173322    411.424483   
1  5.422446  205.266080  18542.957451     5.491963  306.702227    382.080129   
2  7.341547  187.672402  21273.457066     7.784003         NaN    332.084293   
3  9.056245  197.666301  17403.532167     7.688917  337.460176    414.766631   
4  5.039374  142.860598  40829.353167     7.271543         NaN    386.803057   

   Organic_carbon  Trihalomethanes  Turbidity  Potability  
0       19.585002        25.057375   4.028958           0  
1       10.504023        67.493450   2.911751           1  
2       16.842334        55.019151   4.025644           0  
3       15.349869        63.696746   3.319354           0  
4       16.823773        52.297113   4.957420           0

Question 93

ph값은 상당히 많은 결측치를 포함한다. 결측치를 제외한 나머지 데이터들 중 사분위값 기준 하위 25%의 값들의 평균값은?

In [212]:
# a = df.ph.dropna()
a = df[df['ph'].notnull()].reset_index()
a[a['ph'] <= a['ph'].quantile(0.25)].ph.mean()

# 5.057093462441731

5.057093462441731

# 의료 비용 예측 데이터

데이터 설명 : 의료비용 예측문제

In [217]:
import pandas as pd
train = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/MedicalCost/train.csv")
train.head()

age     sex     bmi  children smoker     region      charges
0   64  female  39.330         0     no  northeast  14901.51670
1   47  female  27.830         0    yes  southeast  23065.42070
2   52  female  33.300         2     no  southwest  10806.83900
3   33  female  22.135         1     no  northeast   5354.07465
4   30    male  31.400         1     no  southwest   3659.34600

Question 94

흡연자와 비흡연자 각각 charges의 상위 10% 그룹의 평균의 차이는?

In [219]:
train.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')

In [221]:
smoke = train[train['smoker'] == 'yes']
smoke = smoke[smoke['charges'] >= smoke['charges'].quantile(0.9)].charges.mean()

not_smoke = train[train['smoker'] == 'no']
not_smoke = not_smoke[not_smoke['charges'] >= not_smoke['charges'].quantile(0.9)].charges.mean()

smoke - not_smoke

# 29297.954548156144

29297.954548156144

# 킹카운티 주거지 가격예측문제 데이터

데이터 설명 : 킹카운티 주거지 가격 예측문제

In [222]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/kingcountyprice//train.csv")
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  8961990160  20150413T000000  567500.0         3        2.5         2080   
1  9455200205  20140604T000000  525000.0         3        2.0         1540   
2  7853220670  20140918T000000  540000.0         3        2.5         2860   
3  3298201170  20141110T000000  350000.0         3        1.0          940   
4  7972604355  20140521T000000  218000.0         3        1.0         1020   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      4556     2.0           0     0  ...      8        2080              0   
1      7800     1.0           0     0  ...      8        1540              0   
2      8935     2.0           0     0  ...      8        2860              0   
3      7811     1.0           0     0  ...      6         940              0   
4      7874     1.0           0     0  ...      7        1020              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1999             0    98074  47.6036 -122.014           1530   
1      2004             0    98125  47.7041 -122.288           1510   
2      2004             0    98065  47.5336 -121.855           2650   
3      1959             0    98008  47.6195 -122.118           1180   
4      1956             0    98106  47.5175 -122.346           1290   

   sqft_lot15  
0        5606  
1        7800  
2        6167  
3        7490  
4        7320  

[5 rows x 21 columns]

Question 95

bedrooms 의 빈도가 가장 높은 값을 가지는 데이터들의 price의 상위 10%와 하위 10%값의 차이를 구하여라

In [224]:
max_bed = df['bedrooms'].value_counts().index[0]
df[df['bedrooms'] == max_bed].price.quantile(0.9) - df[df['bedrooms'] == max_bed].price.quantile(0.1)

# 505500.0

505500.0

# 대학원 입학가능성 데이터

데이터 설명 : 대학원 입학 가능성 예측

In [231]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/admission/train.csv")
df.head()

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0         246        328          110                  4  4.0  2.5  9.02   
1         100        323          113                  3  4.0  4.0  8.88   
2          79        296           95                  2  3.0  2.0  7.54   
3          53        334          116                  4  4.0  3.0  8.00   
4         444        321          114                  5  4.5  4.5  9.16   

   Research  Chance of Admit  
0         1             0.81  
1         1             0.79  
2         1             0.44  
3         1             0.78  
4         1             0.87

Question 96

Serial No. 컬럼을 제외하고 ‘Chance of Admit’을 종속변수, 나머지 변수를 독립변수라 할때, 랜덤포레스트를 통해 회귀 예측을 할 떄 변수중요도 값을 출력하라 (시드값에 따라 순서는 달라질수 있음)

In [232]:
from sklearn.ensemble import RandomForestRegressor

df = df.drop('Serial No.',axis=1)
x = df.drop('Chance of Admit',axis=1)
y = df['Chance of Admit']

ml = RandomForestRegressor()
ml.fit(x,y)

RandomForestRegressor()

In [235]:
x.columns

Index(['GRE Score', 'TOEFL Score', 'University Rating', 'SOP', 'LOR', 'CGPA',
       'Research'],
      dtype='object')

In [233]:
ml.feature_importances_

array([0.09185323, 0.03649806, 0.01288519, 0.02469082, 0.02414692,
       0.80201204, 0.00791374])

In [241]:
pd.DataFrame({'importance':ml.feature_importances_},x.columns).sort_values('importance',ascending=False)

importance
CGPA                 0.802012
GRE Score            0.091853
TOEFL Score          0.036498
SOP                  0.024691
LOR                  0.024147
University Rating    0.012885
Research             0.007914

# 레드 와인 퀄리티 예측 데이터

데이터 설명 : 레드 와인 퀄리티 예측문제

In [242]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/redwine/train.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            5.0              0.38         0.01             1.6      0.048   
1            5.0              0.42         0.24             2.0      0.060   
2            7.1              0.36         0.30             1.6      0.080   
3            7.6              0.29         0.49             2.7      0.092   
4            7.7              0.51         0.28             2.1      0.087   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 26.0                  60.0  0.99084  3.70       0.75   
1                 19.0                  50.0  0.99170  3.72       0.74   
2                 35.0                  70.0  0.99693  3.44       0.50   
3                 25.0                  60.0  0.99710  3.31       0.61   
4                 23.0                  54.0  0.99800  3.42       0.74   

   alcohol  quality  
0     14.0        6  
1     14.0        8  
2      9.4        5  
3     10.1        6  
4      9.2        5

Question 97

quality 값이 3인 그룹과 8인 데이터그룹의 각 컬럼별 독립변수의 표준편차 값의 차이를 구할때 그값이 가장 큰 컬럼명을 구하여라

In [249]:
abs(df[df.quality == 8].std() - df[df.quality == 3].std()).sort_values().index[-1]

# total sulfur dioxide

'total sulfur dioxide'

# 약물 분류 데이터

데이터 설명 : 투약하는 약을 분류

In [250]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/drug/train.csv")
df.head()

Age Sex      BP Cholesterol  Na_to_K  Drug
0   20   F  NORMAL      NORMAL    9.281     4
1   24   M    HIGH      NORMAL    9.475     1
2   34   M  NORMAL        HIGH   22.456     0
3   17   M  NORMAL      NORMAL   10.832     4
4   72   M     LOW        HIGH   16.310     0

Question 98

남성들의 연령대별 (10살씩 구분 0~9세 10~19세 …) Na_to_K값의 평균값을 구해서 데이터 프레임으로 표현하여라

In [257]:
df['new_Age'] = df['Age']//10*10
df[df['Sex'] == 'M'].groupby('new_Age')['Na_to_K'].mean().to_frame()

Na_to_K
new_Age           
10       13.627000
20       18.260769
30       15.244143
40       13.897273
50       14.811273
60       14.284308
70       11.363857

# 사기회사 분류 데이터

데이터 설명 : 사기회사 분류

In [258]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/audit/train.csv")
df.head()

Sector_score LOCATION_ID  PARA_A  Score_A  Risk_A  PARA_B  Score_B  Risk_B  \
0          2.72          14    4.53      0.6   2.718   87.93      0.6  52.758   
1          3.41          16   12.68      0.6   7.608   41.00      0.6  24.600   
2          2.72          11    3.11      0.6   1.866  113.97      0.6  68.382   
3          3.41           8    1.12      0.4   0.448    0.00      0.2   0.000   
4         55.57           9    1.06      0.4   0.424    0.00      0.2   0.000   

    TOTAL  numbers  ...  RiSk_E  History  Prob  Risk_F  Score  Inherent_Risk  \
0   92.46      5.0  ...     0.4        0   0.2     0.0    4.0        108.362   
1   53.68      5.5  ...     0.4        0   0.2     0.0    4.4         41.936   
2  117.08      5.5  ...     0.4        0   0.2     0.0    4.4         88.832   
3    1.12      5.0  ...     0.4        0   0.2     0.0    2.2          1.848   
4    1.06      5.0  ...     0.4        0   0.2     0.0    2.2          1.824   

   CONTROL_RISK  Detection_Risk  Audit_Risk  Risk  
0           0.4             0.5     21.6724     1  
1           0.4             0.5      8.3872     1  
2           0.4             0.5     17.7664     1  
3           0.4             0.5      0.3696     0  
4           0.4             0.5      0.3648     0  

[5 rows x 27 columns]

Question 99

데이터의 Risk 값에 따른 score_a와 score_b의 평균값을 구하여라

In [260]:
df[['Risk','Score_A','Score_B']].groupby('Risk').mean()

Score_A   Score_B
Risk                    
0     0.262234  0.225532
1     0.490164  0.444262

# 센서데이터 동작유형 분류 데이터

데이터 설명 : 센서데이터로 동작 유형 분류

In [261]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/muscle/train.csv")
df.head()

motion_0  motion_1  motion_2  motion_3  motion_4  motion_5  motion_6  \
0      -6.0       8.0       5.0       6.0     -12.0     -27.0       5.0   
1     -32.0      -3.0      -4.0      -5.0      -1.0      16.0       0.0   
2      35.0      -2.0      -1.0       1.0     -25.0     -16.0      -3.0   
3     -15.0      -6.0      -3.0      -5.0      27.0      37.0      -2.0   
4      60.0       3.0       0.0      -7.0      -6.0     -10.0      -6.0   

   motion_7  motion_8  motion_9  ...  motion_55  motion_56  motion_57  \
0      24.0      -8.0      -5.0  ...       -1.0       -3.0       -6.0   
1      -5.0      42.0       2.0  ...      -17.0      -12.0       -1.0   
2     -20.0      10.0       1.0  ...        6.0      -65.0        6.0   
3      -2.0      19.0       5.0  ...        2.0       11.0        0.0   
4       7.0     -11.0      -4.0  ...      -12.0       17.0       -3.0   

   motion_58  motion_59  motion_60  motion_61  motion_62  motion_63  pose  
0       -9.0       -7.0       36.0       68.0        3.0       -3.0     1  
1       -1.0        0.0      -26.0      -11.0       -4.0       -9.0     1  
2      -10.0      -31.0      -19.0       -3.0      -47.0      -17.0     0  
3       -2.0       -4.0       -7.0       15.0        3.0       14.0     1  
4        1.0        2.0       -2.0       20.0       18.0       10.0     0  

[5 rows x 65 columns]

Question 100

pose값에 따른 각 motion컬럼의 중간값의 가장 큰 차이를 보이는 motion컬럼은 어디이며 그값은?

In [267]:
a = df.groupby('pose').median().T
abs(a[0] - a[1]).sort_values()

# mothio_54, mothion_62

motion_0     0.0
motion_24    0.0
motion_25    0.0
motion_26    0.0
motion_32    0.0
            ... 
motion_30    3.0
motion_38    3.0
motion_46    3.5
motion_54    4.0
motion_62    4.0
Length: 64, dtype: float64

# 현대 차량 가격 분류문제 데이터

데이터 설명 : 현대 차량가격 분류문제

In [269]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/hyundai/train.csv")
df.head()

model  year  price transmission  mileage fuelType  tax(£)   mpg  \
0   Tucson  2016  12795       Manual    36263   Diesel      30  61.7   
1      I10  2012   3995       Manual    22089   Petrol      20  61.4   
2      I30  2019  24000       Manual      555   Petrol     145  34.0   
3   Tucson  2017  12995       Manual    32147   Diesel      30  61.7   
4   Tucson  2018  14991    Semi-Auto    32217   Diesel     150  57.6   

   engineSize  
0         1.7  
1         1.2  
2         2.0  
3         1.7  
4         1.7

Question 101

정보(row수)가 가장 많은 상위 3차종의 price값의 각 평균값은?

In [275]:
df[df.model.isin(df.model.value_counts().index[:3])].groupby('model').price.mean()

model
 I10        7646.137891
 I30       11651.821759
 Tucson    15805.441373
Name: price, dtype: float64

# 당뇨여부판단 데이터

데이터 설명 : 당뇨여부 판단하기

In [276]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/diabetes/train.csv")
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            3      102             74              0        0  29.5   
1            2      144             58             33      135  31.6   
2            5      136             82              0        0   0.0   
3           13      145             82             19      110  22.2   
4            1      117             60             23      106  33.8   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.121   32        0  
1                     0.422   25        1  
2                     0.640   69        0  
3                     0.245   57        0  
4                     0.466   27        0

Question 102

Outcome 값에 따른 각 그룹의 각 컬럼의 평균 차이를 구하여라

In [280]:
a = df.groupby('Outcome').mean().T
abs(a[0] - a[1])

Pregnancies                  1.574159
Glucose                     29.943995
BloodPressure                3.202079
SkinThickness                2.952033
Insulin                     33.398645
BMI                          4.541437
DiabetesPedigreeFunction     0.143646
Age                          5.866939
dtype: float64

# 넷플릭스 주식 데이터

데이터 설명 : 넷플릭스 주식데이터

In [281]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nflx/NFLX.csv")
df.head()

Date      Open      High       Low     Close  Adj Close     Volume
0  2002-05-23  1.156429  1.242857  1.145714  1.196429   1.196429  104790000
1  2002-05-24  1.214286  1.225000  1.197143  1.210000   1.210000   11104800
2  2002-05-28  1.213571  1.232143  1.157143  1.157143   1.157143    6609400
3  2002-05-29  1.164286  1.164286  1.085714  1.103571   1.103571    6757800
4  2002-05-30  1.107857  1.107857  1.071429  1.071429   1.071429   10154200

Question 103

매년 5월달의 open가격의 평균값을 데이터 프레임으로 표현하라

# nba 선수 능력치 데이터

Question 104

Tm 컬럼은 각 팀의 이름을 의미한다. TOR팀의 평균나이를 소수 4째 자리까지 구하여라

Question 105

Pos 컬럼은 포지션을 의미한다. 전체 선수 중 최소나이대의 선수들을 필터하고 그들 중 가장 많은 포지션은 무엇인지 확인하라

Question 106

선수들의 이름은 first_name+ 공백 + last_name으로 이루어져 있다. 가장 많은 first_name은 무엇이며 몇 회 발생하는지 확인하라

Question 107

PTS컬럼은 경기당 평균득점수 이다. 각포지션별로 경기당 평균득점수의 평균을 구하여라

Question 108

PTS컬럼은 경기당 평균득점수 이다. 각포지션별로 경기당 평균득점수의 평균을 구하여라

Question 109

G컬럼은 참여한 경기의 숫자이다. 각 팀별로 가장 높은 경기참여 수를 가진 선수들의 경기 참여 숫자의 평균을 구하여라

Question 110

Tm의 값이 MIA이며 Pos는 C또는 PF인 선수의 MP값의 평균은?

Question 111

전체 데이터중 G의 평균값의 1.5배 이상인 데이터들만 추출했을때 3P값의 평균은?

Question 112

Age의 평균 이상인 그룹과 평균 미만인 그룹간의 G값의 평균의 차이는?

Question 113

평균나이가 가장 젊은 팀은 어디인가

Question 114

Pos그룹별 평균 MP값을 구하여라